File: MyIM/MyIM/IM/Cells/MIMMessageTableCell.m
//
//  MIMMessageTableCell.m
//  MyIM
//
//  Created by Jonathan on 15/8/12.
//  Copyright (c) 2015年 Jonathan. All rights reserved.
//

#import "MIMMessageTableCell.h"

#import "UIButton+WebCache.h"

@interface MIMMessageTableCell ()

@property (assign, nonatomic) CGSize          avatarSize;

@property (weak, nonatomic) IBOutlet UILabel *timeLabel;
@property (weak, nonatomic) IBOutlet UIButton *avatarButton;
@property (weak, nonatomic) IBOutlet UILabel *nicknameLabel;


@property (weak, nonatomic) IBOutlet NSLayoutConstraint *avatarHeightConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *avatarWidthConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *timeHeightConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *nicknameHeightConstraint;

@property (strong, nonatomic) NSLayoutConstraint *contentMHC;
@property (strong, nonatomic) NSLayoutConstraint *contentMVC;

@property (strong, nonatomic) UIButton      *errorButton;

@end

@implementation MIMMessageTableCell
@synthesize reuseIdentifier = _reuseIdentifier;
- (instancetype)initWithCellStyle:(MIMMessageCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    NSString *nibName = style == MIMMessageCellStyleIncoming?@"MIMMessageInCell":@"MIMMessageOutCell";
    self = [[[NSBundle mainBundle] loadNibNamed:nibName owner:nil options:nil] lastObject];
    if (self) {
        _style = style;
        _reuseIdentifier = reuseIdentifier;
    }
    return self;
}

- (void)awakeFromNib {

    self.avatarSize         = MIM_AVATAR_SIZE;
    
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}
- (IBAction)avatarClick:(id)sender {
    if (self.avatarClick) {
        self.avatarClick(self.atIndex);
    }
}

#pragma mark - setter -

- (void)setMessageContentView:(UIView *)messageContentView
{
    if (_messageContentView) {
        if (_messageContentView != messageContentView) { //存在 并且和之前的不是同一个view
            [_messageContentView removeFromSuperview];
            [self.contentView addSubview:messageContentView];
            _messageContentView = messageContentView;
        }
        else{
            //存在 并且和之前的是同一个view 不做操作
            return;
        }

    }
    else{
        //不存在
        _messageContentView = messageContentView;
        [self.contentView addSubview:_messageContentView];
    }

    
    [_messageContentView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    //创建约束
        
    NSLayoutConstraint *leadingCt = nil;
    if (self.style == MIMMessageCellStyleOutgoing) {
        leadingCt = [NSLayoutConstraint constraintWithItem:self.avatarButton attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:_messageContentView attribute:NSLayoutAttributeRight multiplier:1.0 constant:MIM_SPACE_BETWEEN_AVATAR];
    }
    else{
        leadingCt = [NSLayoutConstraint constraintWithItem:self.avatarButton attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:_messageContentView attribute:NSLayoutAttributeLeft multiplier:1.0 constant:-MIM_SPACE_BETWEEN_AVATAR];
    }
    
    NSLayoutConstraint *topCt = [NSLayoutConstraint constraintWithItem:self.nicknameLabel attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:_messageContentView attribute:NSLayoutAttributeTop multiplier:1.0 constant:MIM_SPACE_BETWEEN_NICKNAME];
    
    self.contentMHC = [NSLayoutConstraint constraintWithItem:self.messageContentView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:0 multiplier:1.0 constant:0];
    
    self.contentMVC = [NSLayoutConstraint constraintWithItem:self.messageContentView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:0 multiplier:1.0 constant:0];
    
    [self.contentView addConstraints:@[leadingCt, topCt, self.contentMHC, self.contentMVC]];
    [self updateConstraints];

}

- (void)setMessageContentSize:(CGSize)messageContentSize
{
    //与原来相同 不做操作
    if (CGSizeEqualToSize(_messageContentSize, messageContentSize)) {
        return;
    }
    _messageContentSize = messageContentSize;
    
    self.contentMHC.constant = messageContentSize.height;
    self.contentMVC.constant = messageContentSize.width;
    
    [self updateConstraintsIfNeeded];
    
}

- (void)setMessageTime:(NSString *)messageTime
{
    _messageTime = messageTime;
    if (_messageTime) {
        self.timeHeightConstraint.constant = MIM_TIME_LABEL_HEIGHT;
    }
    else{
        self.timeHeightConstraint.constant = 0.0f;
    }
    [self updateConstraintsIfNeeded];
    
    self.timeLabel.text = self.messageTime;

}

- (void)setNickName:(NSString *)nickName
{
    _nickName = nickName;
    if (_nickName) {
        self.nicknameHeightConstraint.constant = MIM_NICKNAME_LABEL_HEIGHT;
    }
    else{
        self.nicknameHeightConstraint.constant = 0.0f;
    }
    [self updateConstraintsIfNeeded];
    
    self.nicknameLabel.text = self.nickName;

}

- (void)setAvatar:(MIMImageModel *)avatar
{
    if (_avatar == avatar) {
        return;
    }
    
    if (avatar) {
        NSString * previousAvatarUrl = _avatar.thumbUrl?_avatar.thumbUrl:_avatar.imageUrl;
        NSString * avatarUrl = avatar?(avatar.thumbUrl?avatar.thumbUrl:avatar.imageUrl):nil;
        //如果存在新url 并且存在前url 并且地址相同 则不做操作，否则重新设置
        if (avatarUrl && previousAvatarUrl
            && [previousAvatarUrl isEqualToString:avatarUrl]) {
        }
        else{
            _avatar = avatar;
            [self.avatarButton sd_setImageWithURL:[NSURL URLWithString:avatarUrl] forState:UIControlStateNormal placeholderImage:self.avatar.placeHolderImage];
        }
        
    }
}

- (void)setShowError:(BOOL)showError
{
    _showError = showError;
    if (showError) {
        [self.contentView bringSubviewToFront:self.errorButton];
        self.errorButton.hidden = NO;
    }
    else{
        self.errorButton.hidden = YES;
    }
}

- (UIButton *)errorButton
{
    if (!_errorButton) {
        _errorButton = [UIButton buttonWithType:UIButtonTypeCustom];
//        _errorButton.frame = CGRectMake(0, 0, 20.0, 20.0);
//        [_errorButton setTitle:@"!" forState:UIControlStateNormal];
        [_errorButton setImage:[UIImage imageNamed:@"message_fail"] forState:UIControlStateNormal];
        [_errorButton addTarget:self action:@selector(errorButtonClick) forControlEvents:UIControlEventTouchUpInside];
        [self.contentView addSubview:_errorButton];
        [_errorButton setTranslatesAutoresizingMaskIntoConstraints:NO];
        
        NSLayoutConstraint *leadingCt = [NSLayoutConstraint constraintWithItem:self.messageContentView attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:_errorButton attribute:NSLayoutAttributeRight multiplier:1.0 constant:3.0f];
        if (self.style == MIMMessageCellStyleIncoming) {
            leadingCt = [NSLayoutConstraint constraintWithItem:self.messageContentView attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:_errorButton attribute:NSLayoutAttributeLeft multiplier:1.0 constant:3.0f];
        }
        
        NSLayoutConstraint *centerCt = [NSLayoutConstraint constraintWithItem:_errorButton attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self.messageContentView attribute:NSLayoutAttributeCenterY multiplier:1.0 constant:0.0f];
        
        NSLayoutConstraint *widthCt  = [NSLayoutConstraint constraintWithItem:_errorButton attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:0 multiplier:1.0 constant:30.0];
        
        NSLayoutConstraint *heightCt  = [NSLayoutConstraint constraintWithItem:_errorButton attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:0 multiplier:1.0 constant:30.0];

        
        [self.contentView addConstraints:@[leadingCt, centerCt, widthCt, heightCt]];
        
    }
    return _errorButton;
}

- (void)errorButtonClick
{
    if(self.errorClick){
        self.errorClick(self.atIndex);
    }
}

@end

File: MyIM/MyIM/Swift/IMCore/Cells/MIMMessageTableCell.swift
//
//  MIMMessageTableCell.swift
//  MyIM
//
//  Created by Jonathan on 16/6/1.
//  Copyright © 2016年 Jonathan. All rights reserved.
//

import UIKit

class MIMMessageTableCell: UITableViewCell {
    
    var style: MIMMessageCellStyle = MIMMessageCellStyle.incoming
    var atIndex: Int = 0
    var messageContentView: UIView?{
        get{
            return self.messageContentView
        }
        set{
            if newValue == nil {
                return
            }
            if self.messageContentView == nil { //view还没有设置过
                if(self.messageContentView !== newValue){
                    self.messageContentView?.removeFromSuperview()
                    self.contentView.addSubview(newValue!)
                    self.messageContentView = newValue
                }
                else{
                    //存在 并且和之前的是同一个view 不做操作
                    return;
                }
            }
            self.updateViewConstraint()
        }
    }
    var messageContentSize: CGSize?{
        get{
            return self.messageContentSize
        }
        set{
            if newValue == nil {
                return;
            }
            if self.messageContentSize != nil && CGSizeEqualToSize(self.messageContentSize!, newValue!){
                return;
            }
            
            self.messageContentSize = newValue
            self.contentMHC!.constant = self.messageContentSize!.height;
            self.contentMVC!.constant = self.messageContentSize!.width;
            self.updateConstraints()
        }
    }
    var avatar: MIMImageItem?{
        get{
            return self.avatar
        }
        set{
            if newValue == nil || self.avatar === newValue {
                return;
            }
            let previousAvatarUrl: String? = self.avatar?.thumbUrl == nil ?self.avatar?.thumbUrl:self.avatar?.url
            let avatarUrl: String? = newValue?.thumbUrl == nil ?newValue?.thumbUrl:newValue?.url
            if previousAvatarUrl != nil && avatarUrl != nil && previousAvatarUrl != avatarUrl {
                
            }
            else{
                self.avatar = newValue
                if avatarUrl == nil {
                    self.avatarButton.setImage(self.avatar?.placeHolderImage, forState: UIControlState.Normal)
                }
                else{
                    self.avatarButton.sd_setImageWithURL(NSURL.init(string: avatarUrl!), forState: UIControlState.Normal, placeholderImage: self.avatar?.placeHolderImage)
                }
            }
        }
    }
    var nickname: String?{
        get{
            return self.nickname
        }
        set{
            self.nickname = newValue
            if self.nickname != nil {
                self.nicknameHeightConstraint.constant = MIMNicknameLabelHeight
            }
            else{
                self.nicknameHeightConstraint.constant = 0.0
            }
            self.nicknameLabel.text = self.nickname
            self.updateConstraintsIfNeeded()
        }
    }
    var messageTime: String?{
        get{
            return self.messageTime
        }
        set{
            self.messageTime = newValue
            if self.messageTime != nil {
                self.timeHeightConstraint.constant = MIMTimeLabelHeight
            }
            else{
                self.timeHeightConstraint.constant = 0
            }
            self.timeLabel.text = self.messageTime
            self.updateConstraintsIfNeeded()
        }
    }
    
    var showError: Bool? {
        get{
            return self.showError
        }
        set{
            if (newValue != nil && newValue!) {
                self.setupErrorButton()
            }
            else{
                self.errorButton?.hidden = true
            }
        }
    }
    
    override var reuseIdentifier: String? {
        get {
            return self.reuseIdentifier
        }
        set{
            self.reuseIdentifier = newValue
        }
    }
    
    @IBOutlet weak private var timeLabel: UILabel!
    
    @IBOutlet weak private var avatarButton: UIButton!
    
    @IBOutlet weak private var nicknameLabel: UILabel!
    
    @IBOutlet weak private var avatarHeightConstraint: NSLayoutConstraint!
    
    @IBOutlet weak private var avatarWidthConstraint: NSLayoutConstraint!
    
    @IBOutlet weak private var timeHeightConstraint: NSLayoutConstraint!
    
    @IBOutlet weak private var nicknameHeightConstraint: NSLayoutConstraint!
    
    private var contentMHC: NSLayoutConstraint?
    
    private var contentMVC: NSLayoutConstraint?

    private var errorButton:UIButton?
    
    private var avatarClick: ((index: Int)->())?
    
    private var errorClick: ((index: Int)->())?


    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
    }

    override func setSelected(selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

    class func nib(style: MIMMessageCellStyle, reuseIdentifier: String) -> MIMMessageTableCell {
        let nibName: String = (style == MIMMessageCellStyle.incoming ?"MIMMessageInCell" : "MIMMessageOutCell")
        let cell: MIMMessageTableCell = (NSBundle.mainBundle().loadNibNamed(nibName, owner: nil, options: nil).first as? MIMMessageTableCell)!
        cell.reuseIdentifier = reuseIdentifier
        cell.style = style
        return cell
    }
    
    
    func updateViewConstraint(){
        self.messageContentView?.translatesAutoresizingMaskIntoConstraints = false
        var leadingCt: NSLayoutConstraint?
        
        if self.style == MIMMessageCellStyle.outgoing {
            leadingCt = NSLayoutConstraint.init(item: self.avatarButton, attribute: NSLayoutAttribute.Left, relatedBy: NSLayoutRelation.Equal, toItem: self.messageContentView, attribute: NSLayoutAttribute.Right, multiplier: 1.0, constant: MIMAvatarAndContentSpace)
        }
        else{
            leadingCt = NSLayoutConstraint.init(item: self.avatarButton, attribute: NSLayoutAttribute.Right, relatedBy: NSLayoutRelation.Equal, toItem: self.messageContentView, attribute: NSLayoutAttribute.Left, multiplier: 1.0, constant: -MIMAvatarAndContentSpace)
        }
        
        let topCt = NSLayoutConstraint.init(item: self.nicknameLabel, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.messageContentView, attribute: NSLayoutAttribute.Top, multiplier: 1.0, constant: MIMNicknameAndContentSpace)
        
        self.contentMHC = NSLayoutConstraint.init(item: self.messageContentView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 0, constant: 0)
        self.contentMVC = NSLayoutConstraint.init(item: self.messageContentView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1.0, constant: 0.0)
        
        self.contentView.addConstraints([leadingCt!, topCt, self.contentMHC!, self.contentMVC!])
        self.updateConstraints()

    }
    
    func setupErrorButton(){
        if self.errorButton == nil {
            self.errorButton = UIButton.init(type: UIButtonType.Custom)
            self.errorButton?.setImage(UIImage.init(named: "message_fail"), forState: UIControlState.Normal)
            self.errorButton?.addTarget(self, action: #selector(errorButtonClick), forControlEvents: UIControlEvents.TouchUpInside)
        }
        
        self.contentView.bringSubviewToFront(self.errorButton!)

    }
    
    func errorButtonClick() {
        if self.errorClick != nil {
            self.errorClick!(index: self.atIndex)                                                                                                                                                                   
        }
    }
    
    @IBAction private func avatarButtonClick(sender: UIButton) {
        if self.avatarClick != nil {
            self.avatarClick!(index: atIndex)
        }
    }
    
    func receiveAvatarClick(click: (index: Int) ->()) {
        self.avatarClick = click
    }
    
    func receiveErrorClick(click: (index: Int) ->()) {
        self.errorClick = click
    }
}
